lib/grandpa/finalization.go

```diff
@@ -0,0 +1,473 @@
+		return fmt.Errorf("%w: %s", errServicesStopFailed, finalizationEngErr)
+	case votingRoundErr != nil:
+		return fmt.Errorf("%w: %s", errServicesStopFailed, votingRoundErr)
+	}
+
+	return nil
+}
+
+func (fh *finalizationHandler) Stop() (err error) {
+	close(fh.stopCh)
+	<-fh.handlerDone
+
+	return fh.stop()
+}
+
+// runEphemeralServices starts the two ephemeral services that handle the
+// votes for the current round, and returns with nil when the two
+// service runs succeed.
+// If any service run fails, the other service run is stopped and
+// an error is returned. The function returns nil is the finalisation
+// handler is stopped.
+func (fh *finalizationHandler) runEphemeralServices() error {
+	fh.servicesLock.Lock()
+	fh.finalizationEngine, fh.votingRound = fh.newServices()
+	fh.servicesLock.Unlock()
+
+	finalizationEngineErr := make(chan error)
+	go func() {
+		finalizationEngineErr <- fh.finalizationEngine.Run()
+	}()
+
+	votingRoundErr := make(chan error)
+	go func() {
+		votingRoundErr <- fh.votingRound.Run()
+	}()
+
+	for {
+		select {
+		case <-fh.stopCh:
+			return nil
+
+		case err := <-votingRoundErr:
+			if err == nil {
+				votingRoundErr = nil
+				// go out from the select case
+				break
+			}
+
+			stopErr := fh.finalizationEngine.Stop()
+			if stopErr != nil {
+				logger.Warnf("stopping finalisation engine: %s", stopErr)
+			}
+			return fmt.Errorf("%w: %s", errHandleVotingRoundFailed, err)
+
+		case err := <-finalizationEngineErr:
+			if err == nil {
+				finalizationEngineErr = nil
+				// go out from the select case
+				break
+			}
+
+			stopErr := fh.votingRound.Stop()
+			if stopErr != nil {
+				logger.Warnf("stopping voting round: %s", stopErr)
+			}
+
+			return fmt.Errorf("%w: %s", errFinalizationEngineFailed, err)
+		}
+
+		finish := votingRoundErr == nil && finalizationEngineErr == nil
+		if finish {
+			return nil
+		}
+	}
+}
+
+// handleVotingRound interacts with finalizationEngine service
+// executing the actions based on what it receives throuhg channel
+type handleVotingRound struct {
+	grandpaService       *Service
+	finalizationEngineCh <-chan engineAction
+	stopCh               chan struct{}
+	engineDone           chan struct{}
+}
+
+func newHandleVotingRound(service *Service, finalizationEngineCh <-chan engineAction) *handleVotingRound {
+	return &handleVotingRound{
+		grandpaService:       service,
+		stopCh:               make(chan struct{}),
+		engineDone:           make(chan struct{}),
+		finalizationEngineCh: finalizationEngineCh,
+	}
+}
+
+func (h *handleVotingRound) Stop() (err error) {
+	close(h.stopCh)
+	<-h.engineDone
+
+	return nil
+}
+
+func (h *handleVotingRound) Run() (err error) {
+	defer func() {
+		close(h.engineDone)
+		h.engineDone = nil
+	}()
+
+	start := time.Now()
+
+	logger.Debugf("starting round %d with set id %d",
+		h.grandpaService.state.round, h.grandpaService.state.setID)
+
+	for {
+		select {
+		case <-h.stopCh:
+			return nil
+
+		case action := <-h.finalizationEngineCh:
```

Comment:
Question, how long can this case take at most (and why ideally)?

Reply:
Ideally, a round should take seconds to complete.

I believe we should keep them as independent routines since inside a round we have many interactions (sending one pre-vote message, waiting for votes, defining the pre-commit block, sending the pre-commit message, waiting for a completable/finalized block, sending the commit message if needed, check if we have received a valid commit message then proceed to the next round) and for some of those actions we should rely on the peer tracking + networking layer and that's the responsibilities there is the voting handler.

Our next step should be to handle the background active voting round (proceed to the next round but the prior round is still available and running waiting for votes) and given the current state we can spin ephemeral services that take care of a specific round context.

Reply:
Right, the fifo order

Reply:
Ok then it makes sense, cool :+1: